### PR TITLE
List support for ytdl-sub dl (#103)

### LINKED ISSUE #103: List support for `ytdl-sub dl`
It assumes all fields are `key.sub_key.value`. Should support lists like `--my_preset.regex.from.title.match._1 'value'`

### PR DESCRIPTION
**Pull request by MariaMozgunova**
Fixes #103

`ytdl-sub dl` now supports the list

### CODE REVIEW

**Comment by MariaMozgunova** [comment]
Is that right to substitute `ArgumentError` error with  `InvalidDlArguments` exception in `DownloadArgsParser` class?

In the current implementation
```
ytdl-sub dl --some.list._n 'value'
```
will not produce a list if `n < 1`. The output will be `{"some": {"list": {"_n": "value"}}`

The following command:
```
ytdl-sub dl --some.list._1 'value' --some.list.foo 'value'
```
will raise `InvalidDlArguments("Cannot specify a single variable to be two different types")` exception.

**Comment by jmbannon** [comment]
> Is that right to substitute `ArgumentError` error with `InvalidDlArguments` exception in `DownloadArgsParser` class?
> 
> In the current implementation
> 
> ```
> ytdl-sub dl --some.list._n 'value'
> ```
> 
> will not produce a list if `n < 1`. The output will be `{"some": {"list": {"_n": "value"}}`
> 
> The following command:
> 
> ```
> ytdl-sub dl --some.list._1 'value' --some.list.foo 'value'
> ```
> 
> will raise `InvalidDlArguments("Cannot specify a single variable to be two different types")` exception.

Hmmm, maybe we should use a different syntax altogether to represent list indices to not have this conflict. Would changing the syntax to `--some.list[1] 'value'` be that hard? I feel bad asking for big changes right after you wrote the code - it looks good!

File: src/ytdl_sub/cli/download_args_parser.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
 import hashlib
+import re
 import shlex
-from argparse import ArgumentError
+from tkinter import N
```
Comment (jmbannon) [inline src/ytdl_sub/cli/download_args_parser.py:4]: Lets get rid of this

File: src/ytdl_sub/cli/download_args_parser.py [hunk 2]
```diff
@@ -116,10 +152,48 @@ def to_subscription_dict(self) -> Dict:
...
+                list_arguments[prefix] = ([], [])
+
+            list_arguments[prefix][0].append(list_index - 1)
+            list_arguments[prefix][1].append(argument_value)
+
+        for name, value in list_arguments.items():
+            n = self._find_largest_consecutive(indices=(value[0]).copy()) + 1
+
+            if n == 0:
+                raise InvalidDlArguments("Incomplete list")
```
Comment (jmbannon) [inline src/ytdl_sub/cli/download_args_parser.py:173]: nit: lets have the `+1` happen inside the function

File: tests/unit/cli/test_download_args_parser.py [hunk 3]
```diff
@@ -66,6 +66,11 @@ class TestDownloadArgsParser:
                 "dl --youtube.playlist_url https://youtube.com/playlist?list=123abc",
                 {"youtube": {"playlist_url": "https://youtube.com/playlist?list=123abc"}},
             ),
+            (
+                None,
+                "dl --parameter.using.list._2 'v2' --parameter.using.list._1 'v1'",
+                {"parameter": {"using": {"list" : ['v1', 'v2']}}},
```
Comment (jmbannon) [inline tests/unit/cli/test_download_args_parser.py:72]: Lets add some more tests, like
- only one list argument
- argument indices in order (_1, _2)
- duplicate indices (_2, _1, _2)

File: src/ytdl_sub/cli/download_args_parser.py [hunk 4]
```diff
@@ -116,10 +152,48 @@ def to_subscription_dict(self) -> Dict:
...
+
+        try:
+            for idx in range(0, len(arguments), 2):
+                argument_name, argument_value = arguments[idx], arguments[idx + 1]
+
+                argument_dict = self._argument_name_and_value_to_dict(
+                    arg_name=argument_name, arg_value=argument_value
+                )
+                mergedeep.merge(subscription_dict, argument_dict, strategy=mergedeep.Strategy.TYPESAFE_REPLACE)
+        except (TypeError):
```
Comment (jmbannon) [inline src/ytdl_sub/cli/download_args_parser.py:195]: what raises the type error? Lets try to make the try-except more narrow - only around the lines that throw it.

File: src/ytdl_sub/cli/download_args_parser.py [hunk 5]
```diff
@@ -116,10 +152,48 @@ def to_subscription_dict(self) -> Dict:
...
+            raise InvalidDlArguments("Cannot specify a single variable to be two different types")
```
Comment (jmbannon) [inline src/ytdl_sub/cli/download_args_parser.py:196]: Lets include the problematic argument name in the error message so its more verbose.
`f"Invalid dl argument {argument_name}: Cannot specify an argument to be two different types"`

**Comment by jmbannon** [comment]
Looks like the linter got you. Run `./tools/linter` and fix any it reports

**Comment by MariaMozgunova** [comment]
Implement all the suggested changes

File: src/ytdl_sub/cli/download_args_parser.py [hunk 6]
```diff
@@ -61,6 +61,46 @@ def _get_argument_name(cls, arg_name: str) -> str:
...
+
+        return bool(pattern.search(argument_name))
+
+    @classmethod
+    def _extract_list_index(cls, argument_name: str) -> int:
+        """
+        :param argument_name: The argument name using list
+        :return: Splits the argument and returns the name and the list index.
+        """
+        prefix, _, last = argument_name.rpartition("[")
```
Comment (jmbannon) [inline src/ytdl_sub/cli/download_args_parser.py:81]: nit: you could use a capture group on the regex above, something like:
`r"\[([1-9]\d*)\]$"` - do the match and extraction in one function

**Comment by jmbannon** [comment]
Thank you again @MariaMozgunova ! Looks great, thanks for your contribution

**Pull request merged**